lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+    for (int i = 0; i < nEv; i++) {
+      if (getVerbosity() >= QUDA_SUMMARIZE)
+        printfQuda("EigValue[%04d]: (%+.16e, %+.16e) residual %.16e\n", i, evals[i].real(), evals[i].imag(), residua[i]);
+    }
+
+    delete r[0];
+    return;
+  }
+
+  EigenSolver::~EigenSolver() {
+    if (tmp1) delete tmp1;
+    if (tmp2) delete tmp2;
+    host_free(residua);
+    host_free(Qmat);
+  }
+  //-----------------------------------------------------------------------------
+  //-----------------------------------------------------------------------------
+
+  // Thick Restarted Lanczos Method constructor
+  TRLM::TRLM(QudaEigParam *eig_param, const DiracMatrix &mat, TimeProfile &profile) :
+    EigenSolver(eig_param, profile),
+    mat(mat)
+  {
+    // Tridiagonal/Arrow matrix
+    alpha = (double*)safe_malloc(nKr*sizeof(double));
+    beta = (double*)safe_malloc(nKr*sizeof(double));
+    for (int i = 0; i < nKr; i++) {
+      alpha[i] = 0.0;
+      beta[i] = 0.0;
+    }
+
+    // Thick restart specific checks
+    if (nKr < nEv + 6) errorQuda("nKr=%d must be greater than nEv+6=%d\n", nKr, nEv + 6);
+
+    if (eig_param->eig_type == QUDA_EIG_LANCZOS
+        && !(eig_param->spectrum == QUDA_SPECTRUM_LR_EIG || eig_param->spectrum == QUDA_SPECTRUM_SR_EIG)) {
+      errorQuda("Only real spectrum type (LR or SR) can be passed to the Lanczos solver");
+    }
+  }
+
+  void TRLM::operator()(std::vector<ColorSpinorField *> &kSpace, std::vector<Complex> &evals)
+  {
+
+    // Check to see if we are loading eigenvectors
+    if (strcmp(eig_param->vec_infile, "") != 0) {
+      loadFromFile(mat, kSpace, evals);
+      return;
+    }
+
+    // Test for an initial guess
+    double norm = sqrt(blas::norm2(*kSpace[0]));
+    if (norm == 0) {
+
+      if (getVerbosity() >= QUDA_SUMMARIZE) printfQuda("Initial residual is zero. Populating with rands.\n");
+      if (kSpace[0]->Location() == QUDA_CPU_FIELD_LOCATION) {
+        kSpace[0]->Source(QUDA_RANDOM_SOURCE);
+      } else {
+        RNG *rng = new RNG(kSpace[0]->Volume(), 1234, kSpace[0]->X());
+        rng->Init();
+        spinorNoise(*kSpace[0], *rng, QUDA_NOISE_UNIFORM);
+        rng->Release();
+        delete rng;
+      }
+    }
+
+    // Normalise initial guess
+    norm = sqrt(blas::norm2(*kSpace[0]));
+    blas::ax(1.0 / norm, *kSpace[0]);
+
+    // Create a device side residual vector by cloning
+    // the kSpace passed to the function.
+    ColorSpinorParam csParamClone(*kSpace[0]);
+    csParam = csParamClone;
+    // Increase Krylov space by one vector
+    kSpace.push_back(ColorSpinorField::Create(csParam));
+    csParam.create = QUDA_ZERO_FIELD_CREATE;
+    r.push_back(ColorSpinorField::Create(csParam));
+    //---------------------------------------------------------------------------
+
+    // Convergence and locking criteria
+    double mat_norm = 0.0;
+    double epsilon = DBL_EPSILON;
+    QudaPrecision prec = kSpace[0]->Precision();
+    if (prec == QUDA_DOUBLE_PRECISION) {
```

Comment:
We should turn this `if` chain into ` switch` statement, and further, we should add `#ifdef` guards that respect the precision `cmake` variable (and throw an error if it's an invalid precision).

Reply:
I'll fix this in my next push.  However, I don't agree with respect to cmake variable since this code doesn't involve template instantiation.